File: 79-word-search/79-word-search.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function(board, word) {
    const n = board.length, m = board[0].length;
    if (word.length < 1) return false;

    const backTrack = (i, j, pos) => {
        if (i === n || i < 0 || j === m || j < 0 || board[i][j] !== word[pos]) return false;
        if (pos === word.length-1) return true;
        board[i][j] = ".";
        const found = 
            backTrack(i+1,j, pos+1) ||
            backTrack(i-1,j, pos+1) ||
            backTrack(i,j+1, pos+1) ||
            backTrack(i,j-1, pos+1);

        board[i][j] = word[pos];
        return found;
    };
    
    for (let i=0;i<n;i++) {
        for (let j=0;j<m;j++) {
            if (board[i][j] === word[0]) {
                const match = backTrack(i,j, 0);
                if (match) return true;
            }
        }
    }

    return false; 
};

File: 79-word-search/79-word-search.cs
public class Solution {
    public bool Exist(char[][] board, string word) {
        int index = 0;
        bool found = false;
        for(int i = 0; i < board.Length; i++) {
            for(int j = 0; j < board[0].Length; j++) {
                if(board[i][j] == word[index]) {
                    found = dfs(board, word, i, j, index);
                    if(found) return true;
                }
            }
        }
        return false;
    }
    
    public bool dfs(char[][] board, string word, int row, int col, int index) {
        if(index == word.Length) {
            return true;
        }
        if(row < 0 || row >= board.Length || col < 0 || col >= board[0].Length) {
            return false;
        }
        if(board[row][col] != word[index]) {
            return false;
        }
        board[row][col] = '.';
        bool found = dfs(board, word, row, col + 1, index + 1) ||
            dfs(board, word, row + 1, col, index + 1) ||
            dfs(board, word, row, col - 1, index + 1) ||
            dfs(board, word, row - 1, col, index + 1);
        board[row][col] = word[index];
        return found;
    }
}
